nexus/db-queries/src/db/datastore/sled.rs

```diff
@@ -288,30 +346,197 @@ impl DataStore {
 
                     define_sql_function!(fn random() -> diesel::sql_types::Float);
 
-                    // We only actually care about one target here, so this
-                    // query should have a `.limit(1)` attached. We fetch all
-                    // sled targets to leave additional debugging information in
-                    // the logs, for now.
+                    // Fetch all viable sled targets
                     let sled_targets = sled_targets
                         .order(random())
                         .get_results_async::<Uuid>(&conn)
                         .await?;
+
                     info!(
                         opctx.log,
-                        "found {} available sled targets", sled_targets.len();
+                        "found {} available sled targets before considering affinity", sled_targets.len();
                         "sled_ids" => ?sled_targets,
                     );
 
-                    if sled_targets.is_empty() {
-                        return Err(err.bail(SledReservationError::NotFound));
+                    let anti_affinity_sleds = lookup_anti_affinity_sleds_query(
+                        instance_id,
+                    ).get_results_async::<(AffinityPolicy, Uuid)>(&conn).await?;
+
+                    let affinity_sleds = lookup_affinity_sleds_query(
+                        instance_id,
+                    ).get_results_async::<(AffinityPolicy, Uuid)>(&conn).await?;
+
+                    // We use the following logic to calculate a desirable sled,
```

Comment:
This is sorta the "meat and potatoes" of this PR -- specifically, updating the logic for instance placement based on affinity groups.

This file also contains several tests for various combinations of affinity groups / instances, and I'm always happy to add more.